### PR TITLE
Remove SSE2-specific logic from UTF-16 validation code

### LINKED ISSUE #51830: Test failure Interop\\StringMarshalling\\UTF8\\UTF8Test\\UTF8Test.cmd
Run: [runtime-coreclr jitstress-isas-x86 20210424.1](https://dev.azure.com/dnceng/public/_build/results?buildId=1106300&view=ms.vss-test-web.build-test-results-tab&runId=33782926&resultId=100333&paneView=debug)

Failed tests:
```
CoreCLR windows x86 Checked jitstress_isas_x86_nosse3 @ Windows.10.Amd64.Open
 -Interop\\StringMarshalling\\UTF8\\UTF8Test\\UTF8Test.cmd

CoreCLR Linux x64 Checked jitstress_isas_x86_nosse41 @ Ubuntu.1804.Amd64.Open
 -Interop/StringMarshalling/UTF8/UTF8Test/UTF8Test.sh
```

**Error message:**
```
Unhandled exception. System.ArgumentException: The output byte buffer is too small to contain the encoded data, encoding 'Unicode (UTF-8)' fallback 'System.Text.EncoderReplacementFallback'. (Parameter 'bytes')
 at System.Text.Encoding.ThrowBytesOverflow() in System.Private.CoreLib.dll:token 0x6003189+0x25
 at System.Text.Encoding.ThrowBytesOverflow(EncoderNLS encoder, Boolean nothingEncoded) in System.Private.CoreLib.dll:token 0x600318a+0x26
 at System.Text.Encoding.GetBytesWithFallback(ReadOnlySpan1 chars, Int32 originalCharsLength, Span1 bytes, Int32 originalBytesLength, EncoderNLS encoder) in System.Private.CoreLib.dll:token 0x600319a+0x16c
 at System.Text.Encoding.GetBytesWithFallback(Char* pOriginalChars, Int32 originalCharCount, Byte* pOriginalBytes, Int32 originalByteCount, Int32 charsConsumedSoFar, Int32 bytesWrittenSoFar) in System.Private.CoreLib.dll:token 0x6003198+0x0
 at System.Text.UTF8Encoding.GetBytes(Char* chars, Int32 charCount, Byte* bytes, Int32 byteCount) in System.Private.CoreLib.dll:token 0x6003331+0x34
 at System.StubHelpers.UTF8Marshaler.ConvertToNative(Int32 flags, String strManaged, IntPtr pNativeBuffer) in System.Private.CoreLib.dll:token 0x60033b4+0x15
 at UTF8StringTests.StringParameterRef(String& s, Int32 index)
 at UTF8StringTests.TestStringPassByRef(String orgString, Int32 index) in UTF8Test.dll:token 0x6000008+0xc
 at Test.Main(String[] args) in UTF8Test.dll:token 0x600001b+0x63

Return code: 1
Raw output file: C:\h\w\B29C09FF\w\AA400981\e\Interop\StringMarshalling\Reports\Interop.StringMarshalling\UTF8\UTF8Test\UTF8Test.output.txt
Raw output:
BEGIN EXECUTION
 "C:\h\w\B29C09FF\p\corerun.exe" UTF8Test.dll 
Expected: 100
Actual: -532462766
END EXECUTION - FAILED
FAILED
Test Harness Exitcode is : 1
To run the test:
> set CORE_ROOT=C:\h\w\B29C09FF\p
> C:\h\w\B29C09FF\w\AA400981\e\Interop\StringMarshalling\UTF8\UTF8Test\UTF8Test.cmd
Expected: True
Actual: False

Stack trace
   at Interop_StringMarshalling._UTF8_UTF8Test_UTF8Test_._UTF8_UTF8Test_UTF8Test_cmd()


```

### PR DESCRIPTION
**Pull request by GrabYourPitchforks**
Resolves https://github.com/dotnet/runtime/issues/51830. Specifically, in the following code paths, we're trying to build up a mask of non-ASCII elements. The SSE41 logic sets the `0x0080` bit for all non-ASCII elements, but the SSE2 logic only sets the `0x0080` bit for elements in the range `[0080 .. 7FFF]`.

https://github.com/dotnet/runtime/blob/11c5181e0d1831c86d9014025e01a54e9f3374fa/src/libraries/System.Private.CoreLib/src/System/Text/Unicode/Utf16Utility.Validation.cs#L119-L132

Normally this would be ok, as the logic later in the method would have noticed the `[8000 .. FFFF]` overflow and manually smeared the relevant bits down to the `0x0080` position. However, the optimization introduced by https://github.com/dotnet/runtime/pull/51671 removed this smear.

### Logic before PR 51671

The smear is in line 143. This forces both the `0x8080` bits to be set if the original unsigned element value is `>= 0x0800`. So even if the original element value was "negative" (`[8000 .. FFFF]`), where _charIsNonAscii_ would be missing its `0x0080` bit set, we'd account for it here.

https://github.com/dotnet/runtime/blob/294a7f9a1b6794d1b467090ae047bc195336aadf/src/libraries/System.Private.CoreLib/src/System/Text/Unicode/Utf16Utility.Validation.cs#L143-L144

### Logic after PR 51671

The smear is removed, which means that we're only guaranteed to set the `0x8000` bit if the original unsigned element value is `>= 0x0800`. The `0x0080` bit would also be set if the original value happened to have that same bit set, but that's not a solid guarantee. So when we later popcnt, we see only `0x8000` and incorrectly report this as a "1 bit set", when we should've instead seen `0x8080` and reported "2 bits set". Since the popcnt is used to calculate the number of UTF-8 bytes required to represent this data, this results in an undercount in the return value of `UTF8Encoding.GetByteCount`.

https://github.com/dotnet/runtime/blob/a56a142b65b322f276281f039032848bc8e51ac0/src/libraries/System.Private.CoreLib/src/System/Text/Unicode/Utf16Utility.Validation.cs#L153-L154

### How this PR resolves the issue

This PR simply removes the SSE2-specific logic from this routine. The SSE41-specific logic remains unchanged. This means that no smearing step is required, since the SSE41 code paths already guarantee that the `0x0080` bit will be set for all elements in the range `[0800 .. FFFF]`. This also means that the perf numbers and codegen provided in https://github.com/dotnet/runtime/pull/51671 remain valid for these architectures. If SSE41 instructions are not available, the code will invoke the shared fallback logic, which already handled these cases correctly and was unaffected by the optimizations in PR 51671.

I've also shored up the unit tests to help cover future regressions here. The unit tests pass on my machine both with SSE41 enabled and with the `COMPLUS_ENABLESSE41=0` env var set.

__Note:__ The bad optimization never shipped in any RTM build or preview release. It only ever appeared in nightlies.

### CODE REVIEW

**Comment by tannergooding** [comment]
@richlander, @jkotas, where are we with https://github.com/dotnet/designs/pull/173 and what ISAs we are going to target by default for the binaries we ship in .NET 6?

Notably this drops the SSE2 codepath and so if we plan on shipping with SSE2 as the default, we may end up with worse perf on startup until this method is rejitted.
However, if we are indeed planning on targeting SSE4.1 or even AVX2 by default then I think this change is fine. SSE4.1 can be emulated on ARM (both Windows on ARM and Apple Silicon) and is available on basically all x86/x64 hardware from 2007 onward.

**Comment by richlander** [comment]
That's an excellent q @tannergooding. It's still on the books as a plan. I've been waiting for @AndyAyersMS and/or @davidwrighton to share where we are at on graceful degradation. And by "waiting", that may mean they already told me and I forgot. Once we have that model in place, then we can test it and ensure that machines that top-top at SSE2 are OK (or OK enough).

Graceful degradation means that SSE2-only machine are opted out of R2R methods that use a newer SIMD instruction set, as opposed to being opted out of whole assemblies. It's the latter that has caused concerns.

**Comment by GrabYourPitchforks** [comment]
Is the "graceful degradation" question the only thing holding this PR up? Or is there some other action that needs to take place before it's ready to merge?

**Review by jkotas — APPROVED** [review]
LGTM.

**Comment by richlander** [comment]
@GrabYourPitchforks -- I don't see anything holding up this PR. Yes, I see it's merged. Just answering the q.

**Pull request merged**
